PLAT-84631: Consolidate item logic

ryanjduffy opened the pull request:
Signed-off-by: Ryan Duffy <ryan.duffy@lge.com>

ryanjduffy commented on styles/variables-gallium.less:
```diff
@@ -174,13 +174,11 @@
+@agate-item-font-size: 40px;
+@agate-item-font-weight: 500;
+@agate-item-label-font-size: 30px;
+@agate-item-side-outset: 32px;
```
This seems a bit large to me. Was this pulled from the specs?

teckliew replied:
Yea 32px is a bit large but it's from here:
<img width="387" alt="itempadding" src="https://user-images.githubusercontent.com/8940009/65540226-cebb1e00-debf-11e9-9f4a-c8ed9aa0957f.png">

ryanjduffy replied:
I see. That seems to be consistent for items other than the menu items.

ryanjduffy commented on Item/Item.module.less:
```diff
@@ -16,13 +63,49 @@
+		}
+
+		&.selected {
+			outline: @agate-item-selected-outline;
```
I noticed a faint gap in Chrome with device emulation when using `outline`. `box-shadow` doesn't seem to produce the same result so it might be a layout-preserving alternative.

![image](https://user-images.githubusercontent.com/788456/65538459-35d6d380-debc-11e9-98a9-ff164aba49de.png)

Djspaceg replied:
Was this gap visible on a non-retina display? Was scaling set to 100% (or something else)?

ryanjduffy replied:
Doesn't seem to repro at 100% scaling.

Djspaceg replied:
My opinion would be to ignore this in that case.

ryanjduffy commented on Item/Item.module.less:
```diff
@@ -4,10 +4,57 @@
+
+			&.labelBefore .label {
+				// temp static values
+				.padding-start-end(0, 8px);
```
Let's move these into variables and remove my comment above.

ryanjduffy commented on Item/Item.module.less:
```diff
@@ -4,10 +4,57 @@
+				box-sizing: border-box;
+				// when showing content and label inline, favor contentt
+				// this should be a variable
+				min-width: 60%;
```
Variable for this and the `max-width` below.

ryanjduffy commented on LabeledItem/LabeledItem.js:
```diff
@@ -9,140 +9,23 @@
+import Item from '../Item';
+
+// eslint-disable-next-line enact/prop-types
+const LabeledItemBase = ({children, titleIcon, ...rest}) => (
```
I forgot we talked about adding warnings so I'll push an update to these shims to address that.

ryanjduffy replied:
Added `deprecate` to each of these.

ryanjduffy commented on Item/Item.js:
```diff
@@ -27,12 +74,36 @@ import Skinnable from '../Skinnable';
+		css: componentCss
+	},
+	computed: {
+		className: ({active, selected, styler}) => styler.append({active, selected})
```
We may want to consider another prop name than `selected` which overlaps with toggle items like `SwitchItem` and `RadioItem`. It's conceivable we define "selectable items" as different than "toggle items" and implement it by consuming the `selected` prop in toggle items. Or maybe we need to implement this in the base `ToggleItem` and the absence of a "toggle indicator" (e.g. a switch or radio item) we use the border.

What do you think @teckliew and @Djspaceg?

teckliew replied:
We can move both `active` and `selected` to ToggleItem. Although, I'm not sure what the distinction is between `active` and `selected`...
<img width="353" alt="toggleitems" src="https://user-images.githubusercontent.com/8940009/65543734-9ff47600-dec6-11e9-9f08-247d750e15af.png">

Djspaceg replied:
Does `Item` `selected` differ in purpose/intent from `ToggleItem` `selected`? Curious if we actually need `selected` `itemSelected` and `active`. That seems like a lot of separate states to #1 maintain and #2 for a developer to correctly understand.

ryanjduffy commented on Item/Item.module.less:
```diff
@@ -16,13 +40,70 @@
+			.itemContent {
+				width: unset;
+
+				&.labelAbove,
```
I've added a guard above for `min-width`/`max-width` which allows us to remove this block since those props won't be set for above/below labels.

Djspaceg replied:
Typically what I've done in the past for these sorts of things is include a class on the base component for this "orientation" scenario and get the checking logic out of javascript, then only apply the min/max-width rules to the horizontal state and min/max-height to the vertical state. This means there are no overrides and no logic in JS determining which is ok.

ryanjduffy commented:
Noting the changes I made:

* Fixed `.label` and `.content` to only apply `max-width` / `min-width` on horizontal layouts
* Added the default `publicClassNames` for item to support overrides for `IconItem`.
* Removed LabeledItem LESS and tests and SlotItem LESS as unused

ryanjduffy commented:
Also still concerned with the inline items. Right now, here's the result:

![image](https://user-images.githubusercontent.com/788456/65640751-ed90e180-df9f-11e9-969d-db58f61b7d51.png)

If we move the `max-width` to the content (rather than the root) and override `display` like this:

```less
	&.inline {
			// overriding ui/Item's inline-block
			display: inline-flex;

			.itemContent {
				max-width: 240px;
			}
	}
```

The result seems better. WDYT?

![image](https://user-images.githubusercontent.com/788456/65640702-d8b44e00-df9f-11e9-9967-f78d74654297.png)

ryanjduffy commented:
Couple more items:
* Added the inline styles I suggested above based on @teckliew's :+1: in slack. Seems to be an improvement which we could revise later as specs solidify
* Added prop types (but not docs) for Item
